test/lit/passes/translate-eh-old-to-new.wast

```diff
@@ -0,0 +1,1425 @@
+    ;; An inner try-delegate targets an outer try whose body type is unreachable
+    ;; (due to a return). In this case we don't need an additional 'br' to an
+    ;; outer block.
+    (try $l0
+      (do
+        (try
+          (do)
+          (delegate $l0)
+        )
+        (return)
+      )
+    )
+  )
+
+  ;; CHECK:      (func $delegate-target-caller-none (type $1)
+  ;; CHECK-NEXT:  (throw_ref
+  ;; CHECK-NEXT:   (block $__binaryen_delegate_caller_target0 (result exnref)
+  ;; CHECK-NEXT:    (call $foo)
+  ;; CHECK-NEXT:    (try_table (catch_all_ref $__binaryen_delegate_caller_target0)
+  ;; CHECK-NEXT:     (call $bar)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:    (call $baz)
+  ;; CHECK-NEXT:    (return)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $delegate-target-caller-none
+    ;; A try-delegate targets the caller whose type is none
+    (call $foo)
+    (try
+      (do
+        (call $bar)
+      )
+      (delegate 0)
+    )
+    (call $baz)
+  )
+
+  ;; CHECK:      (func $multiple-delegates-target-caller-none (type $1)
+  ;; CHECK-NEXT:  (throw_ref
+  ;; CHECK-NEXT:   (block $__binaryen_delegate_caller_target0 (result exnref)
+  ;; CHECK-NEXT:    (call $foo)
+  ;; CHECK-NEXT:    (try_table (catch_all_ref $__binaryen_delegate_caller_target0)
+  ;; CHECK-NEXT:     (call $bar)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:    (try_table (catch_all_ref $__binaryen_delegate_caller_target0)
+  ;; CHECK-NEXT:     (call $bar)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:    (call $baz)
+  ;; CHECK-NEXT:    (return)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $multiple-delegates-target-caller-none
+    ;; Multiple try-delegates target the caller whose type is none
+    (call $foo)
+    (try
+      (do
+        (call $bar)
+      )
+      (delegate 0)
+    )
+    (try
+      (do
+        (call $bar)
+      )
+      (delegate 0)
+    )
+    (call $baz)
+  )
+
+  ;; CHECK:      (func $delegate-target-caller-concrete (type $2) (result i32)
+  ;; CHECK-NEXT:  (throw_ref
+  ;; CHECK-NEXT:   (block $__binaryen_delegate_caller_target0 (result exnref)
+  ;; CHECK-NEXT:    (return
+  ;; CHECK-NEXT:     (block (result i32)
+  ;; CHECK-NEXT:      (call $foo)
+  ;; CHECK-NEXT:      (try_table (result i32) (catch_all_ref $__binaryen_delegate_caller_target0)
+  ;; CHECK-NEXT:       (i32.const 0)
+  ;; CHECK-NEXT:      )
+  ;; CHECK-NEXT:     )
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $delegate-target-caller-concrete (result i32)
+    ;; A try-delegate targets the caller whose type is concrete
+    (call $foo)
+    (try (result i32)
+      (do
+        (i32.const 0)
+      )
+      (delegate 0)
+    )
+  )
+
+  ;; CHECK:      (func $delegate-nested-more (type $1)
+  ;; CHECK-NEXT:  (block $outer3
+  ;; CHECK-NEXT:   (block $catch_all4
+  ;; CHECK-NEXT:    (try_table (catch_all $catch_all4)
+  ;; CHECK-NEXT:     (throw_ref
```

Comment:
Reading testcases like this, I get the impression that the new code is less efficient than the old. Is that right? Specifically it looks like to implement a delegate we do a branch and a throw, and before we didn't have this manual throw operation, so the VM could have just jumped along the delegate path.

If that makes sense, a followup question, would new compiler output be able to avoid the extra overhead?

Reply:
Yeah I also (was aware but realize again) that the old spec is very concise.

I ran the translator on an Adobe binary hanging around somewhere in my local machine and the code size increase is 4.2%. We made the switch knowing this will make the code size worse. But one thing I realized was my estimation was incorrect. As you may know, I once estimated the code size increase (for Adobe) to be around 2.1~2.2%, give that we switch to the option that is similar to what we chose and we remove `delegate`.

The most important case, and actually the only case that matters, a simple `try`-`catch_all`-`rethrow`, which is the most common because it is used for destructors. The generated code for other cases are a lot worse due to tuples and such but they are a lot less frequent so they don't affect the code size much.
```wast
try
  ...
catch_all
  ... ;; cleanup
end
```

My (incorrect) estimation (in the doc) was
```wast
block $label0
  try_table (catch_all_ref $label0)
    ...
  end
end
local.set $exn
... ;; cleanup
local.get $exn
throw_ref
```

But this is missing some instructions. The correct version should be
```wast
block $outer
  block $label0
    try_table (catch_all_ref $label0)
      ...
    end
    br $outer
  end
  local.set $exn
  ... ;; cleanup
  local.get $exn
  throw_ref
end
```
So I missed that we need an outer `block` and `end` and a `br` to it in case exceptions are not thrown. Anyway this seems to make the code size increase double. `delegate` transformation seems to be as I expected it to be, so I don't think that estimation was the culprit...

Anyway, this number might go down after optimizations, but at this point, it's worse than I estimated. Not sure if there is anyway we can mitigate this... If we gzip the binaries the size increase goes down to around 2%.

Reply:
Interesting, thanks for the details!

Personally I feel that 2% or so worse (than the 2% we expected) is not terrible since we are getting advantages in return for this change.

However, the delegate stuff worries me, because my naive expectation is that VMs might emit slower code. Specifically delegates seem like they statically told the VM where to go - "if this try catches it, just go to this other try" - but now we need an extra rethrow - "if this try catches it, rethrow right before the try whose catch I want to enter". The VM would need to pattern-match a rethrow that goes straight into a catch to make that efficient, I guess?

Reply:
I think that's a possibility, but exception code is executed in rare circumstances in the first place, so I wonder if they consist of a significant portion of the runtime. (Desturctors too; `try`-`catch_all` destructors run only when an exception is thrown elsewhere and we need to run destructors before we rethrow the exception to the enclosing scope)

cc @thibaudmichaud ... Would `throw_ref` be slower than `try`-`delegate`, and if so, how much?

Reply:
If this is fast enough in the VM then sorry for the noise here, btw. My understanding of the work the VM does for exceptions is very limited :smile: 

Re-reading the output of this test here (starting on line 1313) I wonder if the delegate which goes to `$l00`, and then throws to be caught immediately again, could avoid the second throw by going directly to `$catch_all4`? We do know this statically so it seems possible. OTOH there is a mismatch because one is a `catch_all` and the other a `catch_all_ref`, but even so, can we jump to a place that does another jump rather than a throw? (Again, all of this is based on my intuition that jumps are faster than throws - if that's inaccurate, feel free to ignore...)

Reply:
That is a good point. `delegate` was indeed just a branch, whereas `throw_ref` calls a runtime function which sets up the exception, triggers the unwinder, etc. which is much slower.
I think we could easily turn this runtime call into a branch in the way that @kripken suggested, i.e. by checking if there is an enclosing try scope in the current function and branching to its handler directly.
That would still be two branches instead of one, but that seems harder to optimise for a much smaller benefit.

Reply:
@kripken I think we can do this in case the targeted `try` has `catch_all`.

Original:
```wast
(try $l0
  (do
    (try
      (do)
      (delegate $l0)
    )
  )
)
```

What we do now:
```wast
(block $outer
  (block $catch_all0
    (try_table (catch_all $catch_all0)
      (throw_ref
        (block $catch_all1 (result exnref)
          (try_table (catch_all_ref $catch_all1)
          )
          (br $outer)
        )
      )
    )
  )
)
```

What you suggested:
```wast
(block $outer
  (block $catch_all0
    (try_table (catch_all $catch_all0)
      (try_table (catch_all_ref $catch_all0)
      )
      (br $outer)
    )
  )
)
```

Did I get what you suggested correctly?

I think this is worth doing not only from the performance perspective, which I worry less about because the exception paths are executed rarely anyway, but for the code size too, given that it can reduce a `throw_ref` and `block`/`end`. But as I said, this is only possible only when the target `try` has `catch_all`. Do you mind if I do it as a follow-up?

Reply:
And about the "correct code" I wrote in https://github.com/WebAssembly/binaryen/pull/6210#discussion_r1446748017, I think actually we can reduce `local.set $exn` and `local.get $exn` pair if we run `local2stack` optimization in `optimize-stack-ir`, which brings down the additional number of instruction to only (?) one more than my previous (incorrect) estimation. We don't support the new EH in the stack IR yet, but adding its support to the stack IR should be simple. Supporting the new EH in the whole optimization pipeline will take more time though.

So this
```wast
block $outer
  block $label0
    try_table (catch_all_ref $label0)
      ...
    end
    br $outer
  end
  local.set $exn
  ... ;; cleanup
  local.get $exn
  throw_ref
end
```

can be
```wast
block $outer
  block $label0
    try_table (catch_all_ref $label0)
      ...
    end
    br $outer
  end
  ... ;; cleanup
  throw_ref
end
```

Reply:
@aheejin 

> Did I get what you suggested correctly?

Yes, exactly.

> Do you mind if I do it as a follow-up?

Sounds good.

> this is only possible only when the target try has catch_all.

To make sure I follow, that is in the example that has `catch_all`, but if it were a catch of a specific tag, that would need to match between the two trys? (But maybe we can optimize in some cases even without a match?)

Reply:
> To make sure I follow, that is in the example that has `catch_all`, but if it were a catch of a specific tag, that would need to match between the two trys? (But maybe we can optimize in some cases even without a match?)

To be more precise, I think this optimization is possible only when the targeted `try` has a single `catch_all`; `try`-`catch`-`catch_all` wouldn't count, given that we don't know whether this exception is gonna be caught in the `catch` or `catch_all`. We can only branch to somewhere when we are certain the exception will be caught there.

I'm not sure what you mean by a match between two `try`s. The inner one is a `try`-`delegate` here, so it can't have tags I think..?

Reply:
Also I think we can optimize an inner `try`-`delegate` targeting an outer `try`-`delegate` too... because the destination for the inner `delegate` should be the same as the destination for the outer `delegate`, wherever that is. In case the outer `delegate` targets another `try`-`catch_all`, the both inner and outer `delegate` can be optimized to target the third `catch_all`... which gets complicated 🫠

Reply:
Right, sorry, I always found the delegate stuff confusing... I was confused because I thought for some reason a delegate was for a particular tag or such. Yeah, sounds like we need a catch-all in the outer one then, as you said.

Btw, one possible option for later optimization work here is to do less in this pass and add general `TryTable` optimizations elsewhere. For example `RemoveUnsedBrs` has a jump-threading section which might be adapted to also thread exceptions perhaps (since exception control flow is just branches, it could fit there, I hope). Anyhow, just a thought for later maybe.

Reply:
I was under the impression that it'd be easier to do things here because this pass has the access to the original `try`-`delegate` structure, which I thought was simpler and didn't involve other instructions such as `block`s. But maybe things are easier to maintain if we use something like jump threading after all..? But even if we use jump threading I guess we should do at least certain things here because jump threading itself wouldn't take care of `throw_ref`s and `catch_all_ref`s.

Reply:
Makes sense, yeah, probably doing some stuff here and some elsewhere is best. This pass should definitely use the delegate-specific info it knows, as you said.

But in general I was thinking that in the long term we will need this pass less, while any general-purpose optimizations on TryTable would be helpful forever. So I'd lean in that direction where possible (but I don't know where that line is).